modules/tensor_mechanics/src/materials/ADComputeSmearedCrackingStress.C

```diff
@@ -0,0 +1,680 @@
+  }
+}
+
+void
+ADComputeSmearedCrackingStress::updateLocalElasticityTensor()
+{
+  const ADReal youngs_modulus =
+      ElasticityTensorTools::getIsotropicYoungsModulus(_elasticity_tensor[_qp]);
+
+  bool cracking_locally_active = false;
+
+  const ADReal cracking_stress = _cracking_stress[_qp];
+
+  if (cracking_stress > 0)
+  {
+    ADRealVectorValue stiffness_ratio_local(1.0, 1.0, 1.0);
+    const ADRankTwoTensor & R = _crack_rotation_old[_qp];
+    ADRankTwoTensor ePrime(_elastic_strain_old[_qp]);
+    ePrime.rotate(R.transpose());
+
+    for (unsigned int i = 0; i < 3; ++i)
+    {
+      // Update elasticity tensor based on crack status of the end of last time step
+      if (_crack_damage_old[_qp](i) > 0.0)
+      {
+        if (_cracking_neg_fraction == 0.0 && MooseUtils::absoluteFuzzyLessThan(ePrime(i, i), 0.0))
+          stiffness_ratio_local(i) = 1.0;
+        else if (_cracking_neg_fraction > 0.0 &&
+                 ePrime(i, i) < _crack_initiation_strain_old[_qp](i) * _cracking_neg_fraction &&
+                 ePrime(i, i) > -_crack_initiation_strain_old[_qp](i) * _cracking_neg_fraction)
+        {
+          const ADReal etr = _cracking_neg_fraction * _crack_initiation_strain_old[_qp](i);
+          const ADReal Eo = cracking_stress / _crack_initiation_strain_old[_qp](i);
+          const ADReal Ec = Eo * (1.0 - _crack_damage_old[_qp](i));
+          const ADReal a = (Ec - Eo) / (4 * etr);
+          const ADReal b = (Ec + Eo) / 2;
+          // Compute the ratio of the current transition stiffness to the original stiffness
+          stiffness_ratio_local(i) = (2.0 * a * etr + b) / Eo;
+          cracking_locally_active = true;
+        }
+        else
+        {
+          stiffness_ratio_local(i) = (1.0 - _crack_damage_old[_qp](i));
+          cracking_locally_active = true;
+        }
+      }
+    }
+
+    if (cracking_locally_active)
+    {
+      // Update the elasticity tensor in the crack coordinate system
+      const bool c0_coupled = MooseUtils::absoluteFuzzyEqual(stiffness_ratio_local(0), 1.0);
+      const bool c1_coupled = MooseUtils::absoluteFuzzyEqual(stiffness_ratio_local(1), 1.0);
+      const bool c2_coupled = MooseUtils::absoluteFuzzyEqual(stiffness_ratio_local(2), 1.0);
+
+      const ADReal c01 = (c0_coupled && c1_coupled ? 1.0 : 0.0);
+      const ADReal c02 = (c0_coupled && c2_coupled ? 1.0 : 0.0);
+      const ADReal c12 = (c1_coupled && c2_coupled ? 1.0 : 0.0);
+
+      const ADReal c01_shear_retention = (c0_coupled && c1_coupled ? 1.0 : _shear_retention_factor);
+      const ADReal c02_shear_retention = (c0_coupled && c2_coupled ? 1.0 : _shear_retention_factor);
+      const ADReal c12_shear_retention = (c1_coupled && c2_coupled ? 1.0 : _shear_retention_factor);
+
+      // Filling with 9 components is sufficient because these are the only nonzero entries
+      // for isotropic or orthotropic materials.
+      std::vector<ADReal> local_elastic(9);
```

Comment:
creating and destroying std::vectors is costly. I would suggest making this a member variable!